src/cryptonote_core/cryptonote_tx_utils.cpp

```diff
@@ -104,6 +95,47 @@ namespace cryptonote
 #endif
     block_reward += fee;
 
+    const bool do_carrot = hard_fork_version >= HF_VERSION_CARROT;
+    if (do_carrot)
+    {
+      try
+      {
+        carrot::CarrotDestinationV1 miner_destination;
+        carrot::make_carrot_main_address_v1(miner_address.m_spend_public_key,
+          miner_address.m_view_public_key,
+          miner_destination);
+
+        const std::vector<carrot::CarrotPaymentProposalV1> normal_payment_proposals = {{
+          .destination = miner_destination,
+          .amount = block_reward,
+          .randomness = carrot::gen_janus_anchor()
+        }};
+
+        std::vector<carrot::CarrotCoinbaseEnoteV1> coinbase_enotes;
+        carrot::get_coinbase_output_enotes(normal_payment_proposals, height, coinbase_enotes);
+
+        tx = carrot::store_carrot_to_coinbase_transaction_v1(coinbase_enotes, extra_nonce);
```

Comment:
Nit: maybe `store_carrot_to_coinbase_transaction_v1` should also explicitly call `tx.invalidate_hashes()` to make sure those are false

Reply:
Nit2: also maybe should make sure all enote block indexes are the same